fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/RecordQueryPlanner.java

```diff
@@ -1424,14 +1427,37 @@ private RecordQueryPlan planScan(@Nonnull CandidateScan candidateScan,
             }
             plan = new RecordQueryScanPlan(possibleTypes, new Type.Any(), candidateScan.planContext.commonPrimaryKey, scanComparisons, candidateScan.reverse, strictlySorted);
         } else {
-            plan = new RecordQueryIndexPlan(candidateScan.index.getName(), candidateScan.planContext.commonPrimaryKey, indexScanComparisons, getConfiguration().getIndexFetchMethod(), candidateScan.reverse, strictlySorted);
+            final FetchIndexRecords fetchIndexRecords = resolveFetchIndexRecords(candidateScan);
+            // If this is a regular fetch using the primary key, we can opt to use the configured fetch method,
+            // if, however, this fetch fetches synthetic constituents, we must do it at the client (at this point).
+            final IndexFetchMethod indexFetchMethod =
+                    fetchIndexRecords == FetchIndexRecords.PRIMARY_KEY
+                    ? getConfiguration().getIndexFetchMethod()
+                    : IndexFetchMethod.SCAN_AND_FETCH;
+            plan = new RecordQueryIndexPlan(candidateScan.index.getName(), candidateScan.planContext.commonPrimaryKey, indexScanComparisons, indexFetchMethod, fetchIndexRecords, candidateScan.reverse, strictlySorted);
             possibleTypes = getPossibleTypes(candidateScan.index);
         }
         // Add a type filter if the query plan might return records of more types than the query specified
         plan = addTypeFilterIfNeeded(candidateScan, plan, possibleTypes);
         return plan;
     }
 
+    @Nonnull
+    private FetchIndexRecords resolveFetchIndexRecords(@Nonnull CandidateScan candidateScan) {
```

Comment:
I still feel like this method needs a bit more explanation in comments in the code, given that it's a little confusing. In particular, noting what empty lists here signifies is probably a good idea. I also am inclined to say that it should handle (even if it means throwing an error) the case where some of the queried record types are synthetic and some of the queried record are not.

Reply:
Added a comment. We also now reject mixed types.